src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java

```diff
@@ -813,7 +813,10 @@ public BaseResponseBean advanceSessionWithEndpoint(MenuSession menuSession, Stri
                 try {
                     screen.init(sessionWrapper);
                     doPostAndSync(menuSession, (FormplayerSyncScreen)screen);
-                    executeAndRebuildSession(menuSession);
+                    Screen nextScreen = menuSession.getNextScreen(false, new EntityScreenContext());
+                    if (nextScreen == null && menuSession.getSessionWrapper().getForm() == null) {
```

Comment:
Maybe it should just check that the op being processed is not the last op in the loop? Because should that be where we want to land up no matter what?